numpy/core/src/multiarray/methods.c

```diff
@@ -1712,11 +1715,9 @@ array_setstate(PyArrayObject *self, PyObject *args)
         return NULL;
     }
     size = PyArray_MultiplyList(dimensions, nd);
-    if (PyArray_DESCR(self)->elsize == 0) {
-        PyErr_SetString(PyExc_ValueError, "Invalid data-type size.");
-        return NULL;
-    }
-    if (size < 0 || size > NPY_MAX_INTP / PyArray_DESCR(self)->elsize) {
+    overflowed = npy_mul_with_overflow_intp(
+        &nbytes, size, PyArray_DESCR(self)->elsize);
```

Comment:
Doesn't `size` need to be >= 0 in order to use `npy_mul_with_overflow_intp`? Seems it should be checked first to avoid a function call.

Reply:
Could do something like
```
invalid = size < 0 || npy_mul_with_overflow_intp(...)
```

Reply:
No need, there's already a explicit `a != 0` case in that function.

There's no point avoiding the function call, because optimizing the performance of empty arrays is pretty pointless

Reply:
Oh I see, the issue is size == -1. Maybe I'll just duplicate the below `if`.